cpp/include/cugraph/visitors/graph_envelope.hpp

```diff
@@ -0,0 +1,206 @@
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Andrei Schaffer, aschaffer@nvidia.com
+//
+
+/**
+ * @brief Set of classes abstracting the type-erasure, templates, and template constraints
+ *        to client code that must supply run-time type information (RTTI) and has no template
+constructs.
+ *
+ *  Goal: be able to call an algorithm (say. louvain() on a type erased graph created from RTTI:
+ * {
+ *  auto graph = make_graph(flags...);
+ *  auto res = louvain(graph, params...);
+ * }
+ * params will be also type-erased (or same type regardless of graph-type); and will
+ * be appropriately passed to the Factory and then converted and passed to Visitor constructor
+*/
+
+#pragma once
+
+#include <functional>
+#include <iostream>
+#include <memory>
+#include <stdexcept>
+#include <type_traits>
+#include <vector>
+
+#include "enum_mapping.hpp"
+#include "graph_enum.hpp"
+
+#include <cugraph/utilities/graph_traits.hpp>
+
+namespace cugraph {
+namespace experimental {
+
+class erased_pack_t;  // forward...
+class return_t;       // forward...
+
+// visitor base, incomplete:
+//
+class visitor_t;  // forward...
+
+// envelope class around all
+// graph classes:
+//
+struct graph_envelope_t {
+  struct base_graph_t {  // necessary to avoid circular dependency
+                         // between graph_base_t and graph_envelope_t
+    virtual ~base_graph_t() {}
+
+    /// virtual void print(void) const = 0;
+
+    virtual void apply(visitor_t& v) const = 0;
+  };
+
+  // abstract factory:
+  //
+  struct visitor_factory_t {
+    virtual std::unique_ptr<visitor_t> make_louvain_visitor(erased_pack_t&) const = 0;
+
+    virtual std::unique_ptr<visitor_t> make_bfs_visitor(erased_pack_t&) const = 0;
+  };
+
+  using pair_uniques_t =
+    std::pair<std::unique_ptr<base_graph_t>, std::unique_ptr<visitor_factory_t>>;
+
+  void apply(visitor_t& v) const
+  {
+    if (p_impl_fact_.first)
+      p_impl_fact_.first->apply(v);
+    else
+      throw std::runtime_error("ERROR: Implementation not allocated.");
+  }
+
+  // void print(void) const
+  // {
+  //   if (p_impl_fact_.first)
+  //     p_impl_fact_.first->print();
+  //   else
+  //     throw std::runtime_error("ERROR: Implementation not allocated.");
+  // }
+
+  std::unique_ptr<base_graph_t> const& graph(void) const { return p_impl_fact_.first; }
+
+  std::unique_ptr<visitor_factory_t> const& factory(void) const { return p_impl_fact_.second; }
+
+  graph_envelope_t(DTypes vertex_tid,
```

Comment:
Question to consider: Does hiding all parameters in the `erased_pack_t` make the code more or less maintainable?

I wonder if we would be better off (from an ease of use perspective) having separate SG and MG constructors for the graph_envelope_t, specifying all non-type-erased parameters explicitly and only including the type erased parameters in the erased_pack_t.

Reply:
The idea for `graph_envelope` is to hide the template params on which the graph depends: `vertex_t, edge_t, weight_t, bool st, bool mg`. By "specifying  all non-type-erased parameters explicitly" you meant their names and number? Or, their types, too? If you meant types then that defeats the purpose of `graph_envelope_t`. 

If you meant just names and number, well that will create problems in the future: as we add new graph types and/or graph constructors, we'd need to do the same for `graph_envelope_t`. That, again, defeats the purpose of an "envelope". 

That being said, I agree we do need some hand-shake agreement / check between what's passed to the `graph_envelope_t` constructor and the correct graph constructor to be dispatched. I will think about that.

Reply:
I appreciate the goal of supporting type erasure.  I guess my thought was that parameter types that don't vary might be more clearly specified explicitly rather than hidden in the `erased_pack_t`.

The problem we're addressing, I believe, is that `vertex_t`, `edge_t` and `weight_t` need to be specified by some calling languages (python being the original motivator) dynamically.  If we look your C++ unit test (in `bfs_test.cpp`), we pass into the graph envelope constructor an erased pack (`ep_graph`) that contains things we need to type erase as well as things we don't really need to type erase.  We certainly *can* type erase everything, as this does.  My question is, do we think we have a *better* API by type erasing everything.  Type erasure eliminates the ability to check types at compile time, which loses something in the code quality.

Specifically, I would be concerned about:
* handle - this is a known type, not sure there's value in type erasing this
* num_vertices - this could be passed as a `size_t` to make it work for any value of vertex_t
* graph_props - this is a structure... I don't imagine we will have different versions of this structure that we want to support
* sorted and check are boolean values... again, don't imagine that we need to type erase those.

Reply:
All of these would need to then be erased in the C-only API level (except maybe the booleans). Also, remember, num_vertices is `vertex_t`. So, you'd need an integer type large enough to cover all `vertex_t` types we support. So, probably `size_t`. Which still is type erasure (because it still needs to be casted to `vertex_t` inside, otherwise you're at th every least allocating too much; most likely would not compile). Not to mention C-only might not have a `size_t`. Etc.

So, since you'd have to erase the types of all these parameters anyway, I'd rather do it in a place where I actually _can_ retrieve their actual types; i.e., in the C++ code.

When calling from the C++ side, we don't need to go through this type-erasure mechanism (I'm only doing it to be able to test it, which is why you see that in my unit test).

Reply:
More importantly, however, is that each algorithm call would require bolierplate code to materialize the type-erasure (specifically, casts). Well, that's exactly what the type-erased pack `ep` does under the hood. Instead of delegating that code to the caller `erased_pack_t` does the actual erasure of types for them, transparently.

Reply:
Again, I'm not disparaging the `erased_pack_t` approach.  This works very well for much of what we're putting through the interface.

My concern is whether *all* things should be type erased in this fashion.  boolean's don't need to be type erased at all.  graph_props probably doesn't need to be type erased at all.  I hadn't thought about the handle class and the C issue... clearly that one must be type erased.  Promoting vertex_t to a size_t and casting back to a smaller size is a common C idiom... although as you observe is also type erasure - just via a different mechanism.

I'm OK going down with this approach.  I'm just raising the question to be sure we're comfortable that we've thought about the implications.  The caller from C will have to meet this API, I'm just questioning whether forcing everything to be type erased (even if they don't technically need it) is the best option for that caller.  Will it be intuitive - with adequate documentation - to call the C-only api, or will it be awkward and confusing even with good documentation?

Perhaps we go with this approach and revisit the decision later when we have more concrete examples.  It's relatively easy to move parameters around if we find it's more intuitive later to structure things differently.

Reply:
Resolving for now.